File: buildSrc/src/main/java/elfeatures/gradle/tasks/TransformJarContentTask.java
package elfeatures.gradle.tasks;

import org.gradle.api.internal.file.copy.CopyAction;
import org.gradle.api.tasks.bundling.Zip;
import org.gradle.work.DisableCachingByDefault;

import javax.inject.Inject;
import java.util.function.UnaryOperator;

@DisableCachingByDefault(because = "Not worth caching")
public abstract class TransformJarContentTask extends Zip {

    private UnaryOperator<String> fileEntryNameTransformer;
    private UnaryOperator<String> dirEntryNameTransformer;

    @Inject
    public TransformJarContentTask() {
        getArchiveExtension().set("jar");
        this.fileEntryNameTransformer = UnaryOperator.identity();
        this.dirEntryNameTransformer = UnaryOperator.identity();
    }

    @Override
    protected CopyAction createCopyAction() {
        return new TransformJarContentCopyAction(
                getArchiveFile().get().getAsFile().toPath(),
                fileEntryNameTransformer,
                dirEntryNameTransformer
        );
    }

    public void fileEntryNameTransformer(UnaryOperator<String> transformer) {
        this.fileEntryNameTransformer = transformer;
    }

    public void dirEntryNameTransformer(UnaryOperator<String> transformer) {
        this.dirEntryNameTransformer = transformer;
    }

}

File: buildSrc/src/main/kotlin/elfeatures/gradle/task/TransformJarContentTask.kt
package elfeatures.gradle.task

import org.gradle.api.internal.file.copy.CopyAction
import org.gradle.api.tasks.bundling.Zip
import org.gradle.work.DisableCachingByDefault
import java.util.function.UnaryOperator
import javax.inject.Inject

@DisableCachingByDefault(because = "Not worth caching")
abstract class TransformJarContentTask @Inject constructor() : Zip() {
    private var fileEntryNameTransformer: UnaryOperator<String>
    private var dirEntryNameTransformer: UnaryOperator<String>

    init {
        archiveExtension.set("jar")
        this.fileEntryNameTransformer = UnaryOperator.identity()
        this.dirEntryNameTransformer = UnaryOperator.identity()
    }

    override fun createCopyAction(): CopyAction {
        return TransformJarContentCopyAction(
            archiveFile.get().asFile.toPath(),
            fileEntryNameTransformer,
            dirEntryNameTransformer
        )
    }

    fun fileEntryNameTransformer(transformer: UnaryOperator<String>) {
        fileEntryNameTransformer = transformer
    }

    fun dirEntryNameTransformer(transformer: UnaryOperator<String>) {
        dirEntryNameTransformer = transformer
    }
}